scripts/simulator/stakingActionsGenerator.js

```diff
@@ -0,0 +1,154 @@
+const BN = web3.utils.BN;
+const BASE = 100; // base for weighted operations
+const { zeroBN, zeroAddress } = require("../../test/helper.js");
+
+//next operation options.
+const DEPOSIT = 'deposit';
+const DELEGATE = 'delegate';
+const WITHDRAW = 'withdraw';
+const NO_ACTION = 'null';
+
+module.exports = { DEPOSIT, DELEGATE, WITHDRAW, NO_ACTION }
+
+module.exports.genNextOp = function genNextOp(loop, numRuns) {
+    let rand = genRandomSeed(7, BASE);
+    let depositWeight;
+    let withdrawWeight;
+    let delegateWeight;
+    // weighted operations
+    // at the start, should have more deposits, then taper off
+    let startRatio = loop / numRuns;
+    if (startRatio < 0.002) {
+        depositWeight = 70;
+        withdrawWeight = 75;
+        delegateWeight = 90;
+    } else {
+        depositWeight = 35;
+        withdrawWeight = 60;
+        delegateWeight = 90;
+    }
+
+    if (rand < depositWeight) return DEPOSIT;
+    if (rand < withdrawWeight) return WITHDRAW;
+    if (rand < delegateWeight) return DELEGATE;
+    return NO_ACTION;
+}
+
+// 97%: amount should be smaller than staker balance
+// 3%: should be bigger than staker balance
+module.exports.genDeposit = async function genDeposit(kncToken, stakers) {
+    let result = {
+        'staker': '',
+        'amount': 0,
+        'msg': '',
+        'isValid': false
+    }
+
+    let rand = genRandomSeed(32, stakers.length);
+    result.staker = stakers[rand];
+    let tokenBal = (await kncToken.balanceOf(result.staker));
+    rand = genRandomSeed(32, BASE);
+    if (rand <= 97) {
+        result.amount = genRandomBN(zeroBN, tokenBal);
+        if (result.amount.eq(zeroBN)) {
+            result.msg = 'invalid deposit: 0 amt',
+            result.isValid = false;
+        } else {
+            result.msg = 'valid deposit';
+            result.isValid = true;
+        }
+        
+    } else {
+        result.amount = genRandomBN(tokenBal.add(new BN(1)), tokenBal.mul(new BN(2)));
+        result.msg = 'invalid deposit';
+        result.isValid = false;
+    }
+    return result;
+}
+
+// 60%: Delegate to another address generated by buidler / ganache
```

Comment:
maybe avoid accurate numbers here.
later if changing number.
have to change comments.

or comment start being wrong.

so maybe just write it has some allocation between these optoins.

Reply:
Kay, TODO tmr =)